pkg/datapath/linux/config/config.go

```diff
@@ -549,6 +555,63 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	return fw.Flush()
 }
 
+// vlanFilterMacros generates VLAN_FILTER macros which
+// are written to node_config.h
+func vlanFilterMacros() (string, error) {
+	devices := make(map[int]bool)
+	for _, device := range option.Config.Devices {
+		ifindex, err := link.GetIfIndex(device)
+		if err != nil {
+			return "", err
+		}
+		devices[int(ifindex)] = true
+	}
+
+	allowedVlans := make(map[int]bool)
+	for _, vlanId := range option.Config.BPFAllowVLAN {
+		allowedVlans[vlanId] = true
+	}
+
+	vlansByIfIndex := make(map[int][]int)
+
+	links, err := netlink.LinkList()
+	if err != nil {
+		return "", err
+	}
+
+	for _, l := range links {
+		vlan, ok := l.(*netlink.Vlan)
+		// if it's vlan device and we're controlling vlan main device
+		// and either all vlans are allowed, or we're controlling vlan device or vlan is explicitly allowed
+		if ok && devices[vlan.ParentIndex] && (allowedVlans[0] || devices[vlan.Index] || allowedVlans[vlan.VlanId]) {
+			vlansByIfIndex[vlan.ParentIndex] = append(vlansByIfIndex[vlan.ParentIndex], vlan.VlanId)
+		}
+	}
+
+	if len(vlansByIfIndex) == 0 {
+		return "return CTX_ACT_DROP;", nil
```

Comment:
The description of the `--bpf-allow-vlan` flag is "List of explicitly allowed VLAN IDs, zero will allow all VLAN IDs", but won't the `CTX_ACT_DROP` here mean that zero will allow *no* VLAN IDs? i.e. should this be a `CTX_ACT_OK` instead?

Reply:
Bad description. I will correct - `--bpf-allow-vlan 0` means all VLAN IDs.

Reply:
And some notices: `len(vlansByIfIndex) == 0` is true only when there is no any vlan device corresponding to main devices. By default all vlan devices are added if they are controlled by cilium. But there might be case when there will be vlan devices wich are not controlled by cilium - in this case we can pass option - `--bpf-allow-vlan 4000,4001` or allow all vlan devices - `--bpf-allow-vlan 0`